### PR TITLE
Create Codenvy Terminal packaging

### PR DESCRIPTION
**Pull request by mkuznyetsov**
Add Codenvy terminal with activity tracking (CODENVY-454)

### CODE REVIEW

File: lib/websocket-terminal/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+## About Eclipse Che
+Eclipse Che is a next generation Eclipse IDE and open source alternative to IntelliJ. This repository is licensed under the Eclipse Public License 1.0. Visit [Eclipse Che's Web site](http://eclipse.org/che) for feature information or the main [Che assembly repository](http://github.com/codenvy/che) for a description of all participating repositories.
```
Comment (garagatyi) [inline lib/websocket-terminal/README.md]: Do we need readme about Eclipse Che here?

File: lib/websocket-terminal/pom.xml [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+                            </arguments>
+                            <environmentVariables>
+                                <GOOS>linux</GOOS>
+                                <GOARCH>arm</GOARCH>
+                                <GOARM>5</GOARM>
+                            </environmentVariables>
+                        </configuration>
+                    </execution>
+                    <execution>
+                        <id>compile-pty-arm5</id>
```
Comment (garagatyi) [inline lib/websocket-terminal/pom.xml]: Please use iterator plugin here

File: lib/websocket-terminal/src/assembly/assembly.xml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    <includeBaseDirectory>true</includeBaseDirectory>
+    <baseDirectory>terminal</baseDirectory>
+    <formats>
+        <format>zip</format>
+    </formats>
+    <fileSets>
+        <fileSet>
+            <directory>${project.build.directory}/${item}/go</directory>
+            <includes>
+                <include>che-websocket-terminal</include>
```
Comment (garagatyi) [inline lib/websocket-terminal/src/assembly/assembly.xml]: Why che if  it is in Codenvy?

**Comment by garagatyi** [comment]
I think it is better to repack Che's artifacts if possible. @skabashnyuk WDYT?

File: lib/websocket-terminal/src/assembly/assembly.xml (hunk 3 above)
Comment (mkuznyetsov) [inline lib/websocket-terminal/src/assembly/assembly.xml]: I decided to keep the same name for this executable across the project

File: lib/websocket-terminal/pom.xml (hunk 2 above)
Comment (mkuznyetsov) [inline lib/websocket-terminal/pom.xml]: used

File: lib/websocket-terminal/README.md (hunk 1 above)
Comment (mkuznyetsov) [inline lib/websocket-terminal/README.md]: removed

File: lib/websocket-terminal/src/main/go/activity.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+func makeActivityRequest() {
+	req, _ := http.NewRequest(http.MethodPut, os.Getenv("CHE_API_ENDPOINT") + "/activity/" + os.Getenv("CHE_WORKSPACE_ID"), nil)
+	client := &http.Client{}
+	_, err := client.Do(req)
+
+	if err != nil {
+		log.Printf("Failed to notify user activity in terminal: %s\n", err)
+	}
+
+}
```
Comment (skabashnyuk) [inline lib/websocket-terminal/src/main/go/activity.go]: new line

**Comment by skabashnyuk** [comment]
ok

File: lib/websocket-terminal/src/main/go/activity.go (hunk 4 above)
Comment (mkuznyetsov) [inline lib/websocket-terminal/src/main/go/activity.go]: ok

File: lib/websocket-terminal/src/main/go/activity.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"time"
+	"log"
+	"net/http"
+	"os"
+)
+
+const threshold int64 = 60
```
Comment (garagatyi) [inline lib/websocket-terminal/src/main/go/activity.go]: Can you make the purpose of variable more obvious by naming or adding comments

**Comment by garagatyi** [comment]
As far as I understand you fire HTTP request synchronously on data retrieval. If activity service responds slowly your terminal won't work well. Please fix that.

File: lib/websocket-terminal/src/main/go/activity.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	for _ = range ticker.C {
+		if wa.active {
+			makeActivityRequest()
+			wa.active = false
+		}
+	}
+}
+
+func makeActivityRequest() {
+	req, _ := http.NewRequest(http.MethodPut, os.Getenv("CHE_API_ENDPOINT") + "/activity/" + os.Getenv("CHE_WORKSPACE_ID"), nil)
```
Comment (garagatyi) [inline lib/websocket-terminal/src/main/go/activity.go]: you can store `os.Getenv("CHE_WORKSPACE_ID")` into variable

Comment (garagatyi) [inline lib/websocket-terminal/src/main/go/activity.go]: and os.Getenv("CHE_API_ENDPOINT")

**Comment by garagatyi** [comment]
@sleshchenko @akorneta @evoevodin

File: lib/websocket-terminal/src/main/go/activity.go (hunk 5 above)
Comment (evoevodin) [inline lib/websocket-terminal/src/main/go/activity.go]: +1

**Comment by evoevodin** [comment]
ok

File: lib/websocket-terminal/.gitignore [hunk 7]
```diff
@@ -0,0 +1,6 @@
+websocket-terminal
```
Comment (sleshchenko) [inline lib/websocket-terminal/.gitignore]: Check that we need this line here

Comment (garagatyi) [inline lib/websocket-terminal/.gitignore]: I agree that we don't need it. If you tried to skip binary file then problem is that default build `go build` will produce binary with another name

**Comment by sleshchenko** [comment]
LGTM

File: lib/websocket-terminal/src/main/go/server.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+	env := os.Environ()
+	env = append(env, "TERM=xterm")
+	wp.Cmd.Env = env
+	wp.Pty, err = pty.Start(wp.Cmd)
+	if err != nil {
+		log.Fatalf("Failed to start command: %s\n", err)
+	}
+	//Set the size of the pty
+	pty.Setsize(wp.Pty, 60, 200)
+	go activity.StartTracking();
```
Comment (garagatyi) [inline lib/websocket-terminal/src/main/go/server.go]: Are you sure that we need activity checker thread per websocket terminal?

**Comment by garagatyi** [comment]
LGTM

**Pull request merged**